src/jit/codegencommon.cpp

```diff
@@ -3429,6 +3431,19 @@ void CodeGen::genReportEH()
 
         CORINFO_EH_CLAUSE_FLAGS flags = ToCORINFO_EH_CLAUSE_FLAGS(HBtab->ebdHandlerType);
 
+        if (isCoreRTABI && (XTnum > 0))
+        {
+            // For CoreRT, CORINFO_EH_CLAUSE_SAMETRY flag means that the current clause covers same
+            // try block as the previous one. The runtime cannot reliably infer this information from
+            // native code offsets because of different try blocks can have same offsets. Alternative
+            // solution to this problem would be inserting extra nops to ensure that different try
+            // blocks have different offsets.
+            if (EHblkDsc::ebdIsSameTry(HBtab, HBtab - 1))
+            {
+                flags = (CORINFO_EH_CLAUSE_FLAGS)(flags | CORINFO_EH_CLAUSE_SAMETRY);
```

Comment:
The SAMETRY bit should only be set on catch clauses, not finally/fault/filter/filter-handler. You could probably assert that here. This is ensured in IL, where only 'catch' is allowed to be "mutually-protect". E.g., the C# `try {} catch {} catch {} finally {}` actually exists in IL as `try { try {} catch {} catch {} } finally {}`. Our EH normalization code and optimizations ensure that the two distinct `try` blocks here get different BasicBlocks, so ebdIsSameTry() won't ever return true.

However, I can't find anything that guarantees that the native code offsets for these two `try` regions don't end up identical, if the outer 'try' is otherwise empty (so we have empty head and tail BasicBlocks for it). I think the reason they end up with different native regions is because we have to generate code to call the finally block.

Maybe on the VM side (we could probably do it here as well), we should assert that if there are identical 'try' regions then the 2nd and subsequent ones have this new bit?